include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+#pragma once
+
+#include <mutex>
+#include <shared_mutex>
+
+#include <rclcpp>
+
+namespace rosparam_shortcuts {
+
+/*
+ * Thread-safe container for generic parameter values, providing callback functions.
+ * This class implements a tree structure of parameters which are declared recursively.
+ * All leafs are of type Parameter<rclcpp::Parameter> which are used for interacting with
+ * the rclcpp node parameter interface using the NodeConfig class further down.
+ */
+template <typename ParameterT>
+class Parameter
+{
+public:
+	Parameter(const std::string& name);
+	Parameter(const std::string& name, const ParameterT& default_value);
+
+	// Fields
+	const std::string& name() const noexcept;
+	std::string description() const noexcept;
+	std::string additional_constraints() const noexcept;
+	bool isNodeParameter() const noexcept;  // true if this parameter contains an rclcpp::Parameter
+
+	// Properties
+	Parameter<ParameterT>& describe(const std::string& description, const std::string& additional_constraints = "");
+	// Allow setting additional properties. i.e. read_only, dynamic_typing?
+	// void setProperties(...);
+
+	// Validity Status
+	const bool isValid() const noexcept;  // errors empty?
+	const bool hasErrors() const noexcept;
+	const bool hasWarnings() const noexcept;
+	// TODO: Insert additional thread-safe access to errors/warnings
+
+	// Read/write-locked data access using std::shared_mutex
+	// See: https://www.youtube.com/watch?v=ozOgzlxIsdg
+	ReadLockedPtr<ParameterT> operator->const;
+	WriteLockedPtr<ParameterT> operator->;
```

Comment:
> Are you sure you actually need a `shared_mutex` here instead of a simple one?

I think it has many advantages, but a simple one might work as well.

> The current API design only locks readers while it copies the values of parameters for the method return value, right?

What do you mean with "only locks readers"? The `ReadLockedPtr` is a scoped lock that prevents write access. The read lock is shared so that multiple callers can access the value (const ref) at the same time. The `WriteLockedPtr` is also a scoped lock, but it's exclusive and allows write access (non-const ref).
Then there are two extra convenience operators which allow copy-get (`operator()`), copy-set (`operator=`) which would use read/write locks internally.

> The concept of a shared mutex really only offers better performance with long locking times for readers and rare write updates.

I would expect that we have many many read calls and only some (declare/update callbacks) write calls.
The `shared_mutex` would allow many readers at the same time which would make it more efficient than a simple one as multiple threads don't need to be synchronized for reading.

Reply:
> I would expect that we have many many read calls

True. That is quite a difference to ROS, where we fetch parameters only rarely via API.
Assuming you want to keep parallel access to the dynamic datastructures possible, the shared read-lock makes sense.